### PR TITLE
Add device_of overload for optional<Tensor>

### PR DESCRIPTION
**Pull request by wenleix**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54263 Migrate kernels with Tensor? to C10 full dispatcher
* **#54262 Add device_of overload for optional<Tensor>**

register_dispatch_key.py might generate device_of call over
optional<Tensor> if it happened to be the first Tensor-like
argument.

Differential Revision: [D27164093](https://our.internmc.facebook.com/intern/diff/D27164093/)

### CODE REVIEW

File: tools/codegen/dest/register_dispatch_key.py [hunk 1]
```diff
@@ -121,7 +121,14 @@ def generate_defn(cpp_sig: CppSignature) -> str:
         elif self.target is Target.ANONYMOUS_DEFINITION:
             impl_name = f"at::native::{f.dispatch[self.dispatch_key]}"
 
-            args_exprs_str = ', '.join(a.name for a in args)
+            args_exprs = []
+            for a in args:
+                # TODO: this is a big HACK, figure out better way...
+                if 'legacy' in impl_name and 'optional<Tensor>' in a.type:
```
Comment (wenleix) [inline tools/codegen/dest/register_dispatch_key.py:127]: cc @smessmer  about current hack.

File: tools/codegen/dest/register_dispatch_key.py [hunk 2]
```diff
@@ -161,9 +166,13 @@ def generate_defn(cpp_sig: CppSignature) -> str:
     globalContext().lazyInitCUDA();
     {cuda_guard_from_tensor_options}
 """
-                elif f.device_guard and device_of is not None:
+                elif f.device_guard and tensor_arg_name is not None:
```
Comment (wenleix) [inline tools/codegen/dest/register_dispatch_key.py:169]: another approach is to still use the first `tesnor_like` parameter, and use `device_of` or `device_or_default` based on whether it's optional.

Comment (wenleix) [inline tools/codegen/dest/register_dispatch_key.py:169]: An alternative approach would be still use the first tensor-like argument, but generate `device_of` or `device_or_default` depends on whether it's optional or not.

**Comment by ezyang** [comment]
Can you say which schemas are affected by this?

**Comment by ezyang** [comment]
The adjustment to codegen seems overwrought here. Why not just add a `device_of(optional<Tensor>)` overload to ATen/DeviceGuard.h?

**Comment by wenleix** [comment]
@ezyang : 

> Can you say which schemas are affected by this?

`_thnn_differentiable_lstm_cell_backward`. Here is generated `RegisterMath.cpp`:

```c++
std::tuple<Tensor,Tensor,Tensor,Tensor,Tensor> wrapper__thnn_differentiable_lstm_cell_backward(const c10::optional<Tensor> & grad_hy, const c10::optional<Tensor> & grad_cy, const Tensor & input_gates, const Tensor & hidden_gates, const c10::optional<Tensor> & input_bias, const c10::optional<Tensor> & hidden_bias, const Tensor & cx, const Tensor & cy) {
    const OptionalDeviceGuard device_guard(device_of(grad_hy));
    return at::native::_thnn_differentiable_lstm_cell_backward(grad_hy, grad_cy, input_gates, hidden_gates, input_bias, hidden_bias, cx, cy);
}
```

`device_of(grad_hy)` complaints, as `grad_hy` is now a `Optional<Tensor>`

**Comment by ezyang** [comment]
I don't think you should fix this in this PR, but there is probably a latent bug in the generated device guards here. Suppose you call this function with nullopt in first argument. Then it will always skip setting the device guard. But if there are other arguments, we ought to have set the device guard. So really what you need to do is find a non-optional tensor argument, or failing that, find the first non-nullopt tensor argument to test. Don't fix this in this PR.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**